### PR TITLE
[SPARK-45919][CORE][SQL] Use Java 16 `record` to simplify Java class definition

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr uses the `record` keyword introduced by [JEP 395](https://openjdk.org/jeps/395) to simplify Java class definition.

### Why are the changes needed?
Using the new feature introduced in Java 16 to simplify Java class definition.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by LuciferYang** [comment]
The cases that have not been fixed in this PR include:

1. Cases that still require additional constructors after modification, such as `YarnShuffleService.AppId`. 

https://github.com/apache/spark/blob/2cac768fabaa7cad40390b2205dd9c5000011e4c/common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java#L583-L589

In these cases, the modification still requires the retention of constructors, failing to achieve simplification.

```scala
 @JsonCreator
 public AppId(@JsonProperty("appId") String appId) {
   this.appId = appId;
 }
```

2. Classes in the hive-thriftserver module.

**Comment by LuciferYang** [comment]
The modifications to `KeyGroupedPartitioning` and `UnknownPartitioning` caused the mima check to fail, so [d598ee8](https://github.com/apache/spark/pull/43796/commits/d598ee8e24511fe8345fa2e46bf20518c8ea3d61) revert them.

File: common/network-common/src/main/java/org/apache/spark/network/protocol/StreamChunkId.java [hunk 1]
```diff
@@ -26,14 +26,7 @@
 /**
 * Encapsulates a request for a particular chunk of a stream.
 */
-public final class StreamChunkId implements Encodable {
```
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/protocol/StreamChunkId.java:29]: Although this is `public`, this is not a part of our public Java doc, right?

- https://spark.apache.org/docs/latest/api/java/index.html

Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/protocol/StreamChunkId.java:29]: Yes

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
This is a nice syntax in general, @LuciferYang .

However, we cannot use this when the class provides information hiding.

```java
jshell> private static class FieldAccessor { private final String field; FieldAccessor(String field) { this.field = field; } }
|  created class FieldAccessor

jshell> var a = new FieldAccessor("secret")
a ==> FieldAccessor@12edcd21

jshell> a.field
|  Error:
|  field has private access in FieldAccessor
|  a.field
|  ^-----^

jshell> private record FieldAccessor2(String field) {}
|  created record FieldAccessor2

jshell> var b = new FieldAccessor2("secret")
b ==> FieldAccessor2[field=secret]

jshell> b.field()
$5 ==> "secret"
```

Could you remove all changes like the above?

**Comment by LuciferYang** [comment]
@dongjoon-hyun I want to clarify the issue. We don't want to use `record` here because `field` in the original class doesn't provide an Accessor, but since `record` automatically generates an Accessor method, it exposes the information of `field`. So, in similar scenarios, we shouldn't use `record`, right?

**Comment by dongjoon-hyun** [comment]
> @dongjoon-hyun I want to clarify the issue. We don't want to use `record` here because `field` in the original class doesn't provide an Accessor, but since `record` automatically generates an Accessor method, it exposes the information of `field`. So, in similar scenarios, we shouldn't use `record`, right?

Yes, exactly. That's the point.

**Comment by LuciferYang** [comment]
> > @dongjoon-hyun I want to clarify the issue. We don't want to use `record` here because `field` in the original class doesn't provide an Accessor, but since `record` automatically generates an Accessor method, it exposes the information of `field`. So, in similar scenarios, we shouldn't use `record`, right?
> 
> Yes, exactly. That's the point.

Is it ok now?

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java [hunk 2]
```diff
@@ -1612,19 +1612,8 @@ private void verifyMetrics(
...
-    private final ByteBuffer buffer;
-    PushBlock(int shuffleId, int shuffleMergeId, int mapIndex, int reduceId, ByteBuffer buffer) {
-      this.shuffleId = shuffleId;
-      this.shuffleMergeId = shuffleMergeId;
-      this.mapIndex = mapIndex;
-      this.reduceId = reduceId;
-      this.buffer = buffer;
-    }
+  private record PushBlock(
+      int shuffleId, int shuffleMergeId, int mapIndex, int reduceId, ByteBuffer buffer) {
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1616]: These fields should be `private`, shouldn't these?

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java [hunk 3]
```diff
@@ -153,24 +153,6 @@ private static ShuffleServiceMetricsInfo getShuffleServiceMetricsInfoForGenericV
       valueName + " value of " + baseName);
   }
 
-  private static class ShuffleServiceMetricsInfo implements MetricsInfo {
```
Comment (dongjoon-hyun) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:156]: `static class` is equivalent to `record`? Specifically, `static` part.

Comment (LuciferYang) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:156]: Yes, Modifier 'static' is redundant for inner records

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java (hunk 2 above)
Comment (LuciferYang) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1616]: Ok, let me revert this one. The previous commit did not revert it due to this in test case.

Comment (LuciferYang) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java]: done

**Comment by LuciferYang** [comment]
wait https://github.com/apache/spark/pull/43817

**Comment by LuciferYang** [comment]
rebased

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for updates, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
Merged to master.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Aggregation.java [hunk 4]
```diff
@@ -28,16 +28,7 @@
...
-
-  public Aggregation(AggregateFunc[] aggregateExpressions, Expression[] groupByExpressions) {
-    this.aggregateExpressions = aggregateExpressions;
-    this.groupByExpressions = groupByExpressions;
-  }
-
-  public AggregateFunc[] aggregateExpressions() { return aggregateExpressions; }
-
-  public Expression[] groupByExpressions() { return groupByExpressions; }
+public record Aggregation(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Aggregation.java:31]: This is public API, are we sure this doesn't break binary compatibility? @LuciferYang

Comment (LuciferYang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Aggregation.java:31]: I wrote the following test code relying on Spark 3.5.3:

```java
package org.apache.spark.sql.catalyst.expressions;

import org.apache.spark.sql.connector.expressions.aggregate.Aggregation;
import org.apache.spark.sql.connector.expressions.Expression;
import org.apache.spark.sql.connector.expressions.aggregate.CountStar;

import java.util.Arrays;

public class AggregationTest {
    public void testCompatibility() {
        CountStar star = new CountStar();
        Aggregation aggregation =
          new Aggregation(new CountStar[]{star}, Expression.EMPTY_EXPRESSION);
        System.out.println(Arrays.toString(aggregation.aggregateExpressions()));
        System.out.println(Arrays.toString(aggregation.groupByExpressions()));
        System.out.println(aggregation.hashCode());
    }
}
```

And I compiled it using Java 8, then invoked this code on 4.0.0-preview1:

```
bin/spark-shell --master local             
WARNING: Using incubator modules: jdk.incubator.vector
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 4.0.0-preview1
      /_/
         
Using Scala version 2.13.14 (OpenJDK 64-Bit Server VM, Java 17.0.12)
Type in expressions to have them evaluated.
Type :help for more information.
Spark context Web UI available at http://192.168.43.26:4040
Spark context available as 'sc' (master = local, app id = local-1727359585773).
Spark session available as 'spark'.

scala> val aggTest = new org.apache.spark.sql.catalyst.expressions.AggregationTest()
val aggTest: org.apache.spark.sql.catalyst.expressions.AggregationTest = org.apache.spark.sql.catalyst.expressions.AggregationTest@7689b31

scala> aggTest.testCompatibility
[COUNT(*)]
[]
-2108932933
```

The old code can be executed in 4.0 without recompilation, so I think this is binary compatible.

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Aggregation.java:31]: ok I just want to double check, thanks for the confirmation!

Comment (LuciferYang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Aggregation.java:31]: OK ~

**Pull request closed without merging**